Tests/Rewriters/ConstraintsRewriterTests.cs

```diff
@@ -0,0 +1,143 @@
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ == 2));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.Equal (2) & Rhino.Mocks.Constraints.Is.Equal (1)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ == 2 && _ == 1));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.Equal (2) | Rhino.Mocks.Constraints.Is.Equal (1)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ == 2 || _ == 1));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.Equal (2) & Rhino.Mocks.Constraints.Is.Equal (1) | Rhino.Mocks.Constraints.Is.Equal (3)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ == 2 && _ == 1 || _ == 3));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.NotEqual (2)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ != 2));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.Same (2)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ == 2));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.NotSame (2)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ != 2));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.GreaterThan (2)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ > 2));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.GreaterThanOrEqual (2)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ >= 2));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.LessThan (2)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ < 2));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.LessThanOrEqual (2)));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ <= 2));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.Null()));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ == null));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (Rhino.Mocks.Constraints.Is.NotNull()));",
+        //language=C#
+        @"_mock.Do (Arg<int>.Matches (_ => _ != null));")]
+    [TestCase (
+        //language=C#
+        @"_mock.Do (Arg<ITestInterface>.Matches (Rhino.Mocks.Constraints.Property.Value (""A"", 42)));",
+        //language=C#
+        @"_mock.Do (Arg<ITestInterface>.Matches (_ => _.A == 42));")]
```

Comment:
I don't quite understand what you mean, since in such a case a NullReferenceException is always thrown.

Reply:
As far as the API goes, `Rhino.Mocks.Constraints.Property.Value (""A"", 42))` does not make it clear, that this will result in a NullReferenceException. Did you verify this behavior in Rhino.Mocks?

Reply:
ok I guess I was wrong, RhinoMocks does not throw an exception if the object is null, I think I have to add an additional null check in the moq version

Reply:
Yep. as I said, probably make it an Elvis? I'm guessing Rhino.Mocks does programatic null-propagation in this case?

Reply:
If the object the constraint is evaluated on is `null` the constraint returns `false`, same thing if the property does not exist or cannot be read.

Reply:
@derlistigelurch 
While browsing the source code, I found some more constraints which should be easy to integrate:
- `PublicField.Value`
- `PublicField.ValueConstraint`
- `PublicField.IsNull`
- `PublicField.IsNotNull`
- `Property.ValueConstraint`
- `Property.IsNull`
- `Property.IsNotNull`

Please refer to the following file for a list of all constraints (Check type usages for the related factory methods): https://github.com/rubicon-oss/CoreRhinoMocks/blob/e84578192e2bc42670164f1c55a8b6217219525f/Rhino.Mocks/Constraints/Constraints.cs#L40

Reply:
@MichaelKetting One thing came to my mind. The `Property.X` & `PublicField.X` constraints can be used to assert fields/properties available on the object but unaccessible due to the current type (e.g. explicitly implemented interfaces). I guess that's the primary usecase too. This would mean that the related rewriter should be reworked, is there an equivalent in Moq @derlistigelurch ?

Reply:
@iam-mholle @derlistigelurch 
Okay, so `_=> _!=null && object.Equals (_.Property, expected)` ?

As for the propery functions: we should not go overboard. if they're not needed right now, I would not implement them unless it's just an hour or so or we have time at the end and think it's a good idea.

Reply:
> As for the propery functions: we should not go overboard. 

that's the answer for the additional constraits right? What about the point about inaccessible properties/fields (https://github.com/rubicon-oss/RhinoMocksToMoqRewriter/pull/20#discussion_r618220420)?

Reply:
Do we actually have those? In case it's just a handful, we can fix those manually and just let the visibility compiler errors after the automatic compilation. Otherwise, PrivateInvoke would be the way to go.

Reply:
> Okay, so _=> _!=null && object.Equals (_.Property, expected) ?

This would be equivalent to `object.Equals (_?.Property, expected)`, since it only returns `true` if `_ != null` and `_.Property == expected` or is there any hidden functionality that I don't know?

Reply:
> > Okay, so _=> _!=null && object.Equals (_.Property, expected) ?
> 
> This would be equivalent to `object.Equals (_?.Property, expected)`, since it only returns `true` if `_ != null` and `_.Property == expected` or is there any hidden functionality that I don't know?

If you would want to check for a `null` value on a property, you could then have the case that the object itself is `null` and the constraint would return `true`.

Reply:
So no, that would not be equivalent.